contract.sol

```diff
@@ -0,0 +1,152 @@
+            rewardTokens.push(_rewardTokens[i]);
+            isRewardToken[_rewardTokens[i]] = true;
+        }
+    }
+
+    // Function to stake tokens
+    function stake(uint256 amount, address rewardToken) external nonReentrant whenNotPaused {
+        if (amount <= 0) revert InvalidAmount();  // Ensure amount is positive
+        if (!isRewardToken[rewardToken]) revert InvalidRewardToken();  // Ensure reward token is valid
+
+        stakingToken.transferFrom(msg.sender, address(this), amount);  // Transfer staked tokens to the contract
+
+        Stake storage userStake = stakes[msg.sender][rewardToken];  // Access the user's stake for this reward token
+        userStake.amount += amount;  // Update the staked amount
+        userStake.startTime = block.timestamp;  // Record the start time of the stake
+
+        emit Staked(msg.sender, amount, rewardToken);  // Emit a Staked event
+    }
+
+    // Function to withdraw staked tokens
+    function withdraw(uint256 amount, address rewardToken) external nonReentrant whenNotPaused {
+        Stake storage userStake = stakes[msg.sender][rewardToken];  // Access the user's stake for this reward token
+        if (userStake.amount < amount) revert WithdrawAmountExceedsStake();  // Ensure the user has enough staked
+
+        uint256 penalty = 0;
+        // Check if the withdrawal is before the minimum staking period ends
+        if (block.timestamp < userStake.startTime + minStakingPeriod) {
+            penalty = (amount * earlyWithdrawalPenalty) / 1e20;  // Calculate penalty based on scaled value
+            amount -= penalty;  // Subtract penalty from the amount to be withdrawn
+            accumulatedPenalties += penalty;  // Add penalty to accumulated penalties
+        }
+
+        userStake.amount -= amount;  // Update the user's staked amount
+        stakingToken.transfer(msg.sender, amount);  // Transfer the remaining amount to the user
+
+        emit Withdrawn(msg.sender, amount, penalty, rewardToken);  // Emit a Withdrawn event
+    }
+
+    // Function to claim rewards based on staked tokens
+    function claimReward(address rewardToken) external nonReentrant whenNotPaused {
+        Stake storage userStake = stakes[msg.sender][rewardToken];  // Access the user's stake for this reward token
+        if (userStake.amount == 0) revert NoStakeFound();  // Ensure the user has a stake
+        if (block.timestamp <= userStake.startTime + minStakingPeriod) revert CannotClaimRewardYet();  // Ensure the minimum staking period has passed
+
+        uint256 stakingDuration = block.timestamp - userStake.startTime;  // Calculate the staking duration
+        uint256 rewardAmount = (userStake.amount * rewardRate * stakingDuration) / 1e18;  // Calculate the reward amount
+
+        IERC20(rewardToken).transfer(msg.sender, rewardAmount);  // Transfer the reward tokens to the user
+
+        userStake.startTime = block.timestamp;  // Reset the stake start time
```

Comment:
Maybe imporve this logic, you can do this in bonus part if some time is left. Otherwise this will work.